src/Compilers/Core/Portable/CommandLine/ErrorLogger.cs

```diff
@@ -381,6 +389,60 @@ public string Add(DiagnosticDescriptor descriptor)
                 list.Sort((x, y) => string.CompareOrdinal(x.Key, y.Key));
                 return list;
             }
+
+            /// <summary>
+            /// Compares descriptors by the values that we write to the log and nothing else.
+            /// 
+            /// We cannot use the IEquatable implementation because it includes message format 
+            /// and we don't write it out. It also ignores tags, which we do write.
+            /// </summary>
+            private sealed class Comparer : IEqualityComparer<DiagnosticDescriptor>
+            {
+                public bool Equals(DiagnosticDescriptor x, DiagnosticDescriptor y)
+                {
+                    if (ReferenceEquals(x, y))
+                    {
+                        return true;
+                    }
+
+                    if (ReferenceEquals(x, null) || ReferenceEquals(y, null))
+                    {
+                        return false;
+                    }
+
+                   return (x.Category == y.Category
+                        && x.DefaultSeverity == y.DefaultSeverity
+                        && x.Description.Equals(y.Description)
+                        && x.HelpLinkUri == y.HelpLinkUri
+                        && x.Id == y.Id
+                        && x.IsEnabledByDefault == y.IsEnabledByDefault
+                        && x.Title.Equals(y.Title)
+                        && x.CustomTags.SequenceEqual(y.CustomTags));
+                }
+
+                public int GetHashCode(DiagnosticDescriptor obj)
+                {
+                    if (ReferenceEquals(obj, null))
+                    {
+                        return 0;
+                    }
+
+                    int hash = Hash.Combine(obj.Category.GetHashCode(),
+                        Hash.Combine(obj.DefaultSeverity.GetHashCode(),
+                        Hash.Combine(obj.Description.GetHashCode(),
+                        Hash.Combine(obj.HelpLinkUri.GetHashCode(),
+                        Hash.Combine(obj.Id.GetHashCode(),
+                        Hash.Combine(obj.IsEnabledByDefault.GetHashCode(),
+                        obj.Title.GetHashCode()))))));
```

Comment:
This is a good question. It's copied from the IEquatable that we can't use so I assume not, but I'll double check. I suspect everything null is being rejected or coerced to empty at construction.